### ISSUE TITLE
"bundle" rules

### ISSUE THREAD

**Issue by msxfm** [RequestPolicyContinued/requestpolicy#338]
<a href="https://github.com/alexlehm"><img src="https://avatars.githubusercontent.com/u/156144?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [alexlehm](https://github.com/alexlehm)**
_Saturday Sep 08, 2012 at 08:23 GMT_
_Originally opened as https://github.com/RequestPolicy/requestpolicy/issues/338_

---

I think it would be useful to have bundle rules for sites that use a specific toolkit or are hosted on a specific server that uses the same set of external domains.

E.g. when a site is hosted on blogspot.com but it using its own domain, there are 2 or 3 external domains always used, so I usually have to whitelist all three, it would be nice if this could be matched somehow so that you only have to whitelist one bundle.

**Comment by nodiscc** [comment]
Here is what bundles could look like in the UI:

![bundles](https://cloud.githubusercontent.com/assets/3950390/5982282/d1258510-a8c1-11e4-8b49-298420aba44c.png)

This mechanism could replace the old "allow_embedded" subscription, allowing the user to whitelist/blacklist a whole set of domains in one click. The available actions would be the same as for normal domain-based rules (eg. `Allow requests from *.domain.org to Media (bundle)`). This would also allow managing subscriptions directly from the menu (eg. `Block requests to Trackers (bundle)`). Domains that are not part of a bundle are listed und the the section `Other`.

The disadvantages it has, UI-wise, is that it makes the menu larger vertically (4 more lines in this example). Sorting is also done by number of requests for each bundle instead of domains - except for `Other` that would always be placed over or below other bundles (in this example I placed it below, on top may be better)

Feedback welcome.

**Comment by nodiscc** [comment]
So far I thought of these possible bundles; they could be implemented as subscriptions first:

 * CDNs: akamai, ajax.googleapis.com, cloudflare, bootstrapcdn, cloudfront.net, ...
 * Media: vimeo, youtube, statickflickr, imgur, ytimg, yimg, ...
 * Maps: google maps, openstreetmap, mapquest, ...
 * Fonts: typekit, fonts.googleapis.com, ...
 * Social: disqus, facebook, twitter, addthis, ...
 * Blogs: bp.blogspot.com, blogger.com, blogblog.com, tumblr.com, ...

Feel free to comment if you can think of more

**Comment by myrdd** [comment]
Thanks for your thoughts so far. I'd like to add:

**user bundles & subscription bundles:**
bundles should not only be available for subscriptions, but also for the user.

**several bundles for a domain:**
a domain should not be bound to a bundle, instead a domain could be in several bundles.
→ *example:* `yimg` might be in the bundles "Media" and "Google"

**bundles containing other bundles:**
we might consider bundles containing other bundles as well, e.g. you could make a sub-bundle of "Maps" which is the "Google Maps" bundle, which contains all the domains of Google Maps

**conservative management of subscription bundles:**
if bundles are distributed via subscriptions, I think we should keep name changes at a minimum. If names do not change, the users could override the bundles they've subscribed to, i.e. they could add or remove domains from a bundle just for their local copy.

**rule: "allow requests to itself":**
it might be useful to create rules like *"allow all requests from bundle XY to itself"*. That should be *not* the default.

**user bundles first:**
I think that before we start with subscription bundles we should enable user bundles. Then we might cautiously add subscription bundles.

**Comment by nodiscc** [comment]
Hi @myrdd maybe there's a middle ground? Maybe there's no need to do a distinction between user bundles and subscription bundles.

**Regarding subscriptions vs bundles**, I think the "bundles" could be added to the subscriptions repo (lists of destinations). Take the maps bundle for example. Instead of _only_ having the option to _always_ allow requests to the _Maps bundle_ from the subscriptions page, these subscriptions/bundles could be allowed directly in the menu, on a per-site basis. 

**Regarding bundles based on domains vs based on desired functionality, and sites in multiple bundles** I think the point of _bundles_ is to help users quickly choosing what to allow, based on what they expect to see on the page. Eg. if you expect to see a map when visiting a site, -> _Allow requests to Maps (bundle) from thissite.net_ is the obvious option. For map services integration, most sites only use _one_ map service, having a _Google_ bundle for this has little usefulness. For blogs, if a site uses tumblr or blogspot as a backend, it only uses one of these; clicking on _Allow requests to Blogs (bundle) from thissite.net_ solves the user's problem. etc.

**Regarding rule: "allow all requests from bundle XY to itself":**, in what case could this be useful? 

**Regarding conservative management of subscription bundles** I don't really understand this point, maybe you can clarify.

To sum up, a very usable way to put "bundles" is simply: add the ability to apply subscription rules _on a per-site basis_, from the menu. Users can still _Allow all requests to maps providers_ from the subscriptions settings page, or from a new entry in the menu.

I hope you get the point, what do you think?
